File: src/main/java/org/moinex/ui/dialog/investment/EditTickerPurchaseController.java
/*
 * Filename: EditTickerPurchaseController.java
 * Created on: January 11, 2025
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.ui.dialog.investment;

import jakarta.persistence.EntityNotFoundException;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import javafx.fxml.FXML;
import javafx.stage.Stage;
import lombok.NoArgsConstructor;
import org.moinex.model.Category;
import org.moinex.model.enums.WalletTransactionStatus;
import org.moinex.model.enums.WalletTransactionType;
import org.moinex.model.investment.TickerPurchase;
import org.moinex.model.wallettransaction.Wallet;
import org.moinex.service.CategoryService;
import org.moinex.service.PreferencesService;
import org.moinex.service.TickerService;
import org.moinex.service.wallet.WalletService;
import org.moinex.util.Constants;
import org.moinex.util.UIUtils;
import org.moinex.util.WindowUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;

/** Controller for the Edit Ticker Purchase dialog */
@Controller
@NoArgsConstructor
public final class EditTickerPurchaseController extends BaseTickerTransactionManagement {
    private TickerPurchase purchase = null;

    /**
     * Constructor
     *
     * @param walletService Wallet service
     * @param categoryService Category service
     * @param tickerService Ticker service
     * @param preferencesService I18n service
     * @note This constructor is used for dependency injection
     */
    @Autowired
    public EditTickerPurchaseController(
            WalletService walletService,
            CategoryService categoryService,
            TickerService tickerService,
            PreferencesService preferencesService) {
        super(walletService, categoryService, tickerService, preferencesService);
        this.preferencesService = preferencesService;
        walletTransactionType = WalletTransactionType.EXPENSE;
    }

    public void setPurchase(TickerPurchase p) {
        this.purchase = p;
        tickerNameLabel.setText(
                purchase.getTicker().getName() + " (" + purchase.getTicker().getSymbol() + ")");
        unitPriceField.setText(purchase.getTicker().getCurrentUnitValue().toString());

        setWalletComboBox(purchase.getWalletTransaction().getWallet());

        descriptionField.setText(purchase.getWalletTransaction().getDescription());
        unitPriceField.setText(purchase.getUnitPrice().toString());
        quantityField.setText(purchase.getQuantity().toString());
        statusComboBox.setValue(purchase.getWalletTransaction().getStatus());
        categoryComboBox.setValue(purchase.getWalletTransaction().getCategory());
        transactionDatePicker.setValue(purchase.getWalletTransaction().getDate().toLocalDate());
        includeInAnalysisCheckBox.setSelected(
                purchase.getWalletTransaction().getIncludeInAnalysis());

        totalPriceLabel.setText(
                UIUtils.formatCurrency(purchase.getWalletTransaction().getAmount()));
    }

    @FXML
    @Override
    protected void handleSave() {
        Wallet wallet = walletComboBox.getValue();
        String description = descriptionField.getText();
        WalletTransactionStatus status = statusComboBox.getValue();
        Category category = categoryComboBox.getValue();
        String unitPriceStr = unitPriceField.getText();
        String quantityStr = quantityField.getText();
        LocalDate buyDate = transactionDatePicker.getValue();

        if (wallet == null
                || description == null
                || description.isBlank()
                || status == null
                || category == null
                || unitPriceStr == null
                || unitPriceStr.isBlank()
                || quantityStr == null
                || quantityStr.isBlank()
                || buyDate == null) {
            WindowUtils.showInformationDialog(
                    preferencesService.translate(
                            Constants.TranslationKeys.INVESTMENT_DIALOG_EMPTY_FIELDS_TITLE),
                    preferencesService.translate(
                            Constants.TranslationKeys.INVESTMENT_DIALOG_EMPTY_FIELDS_MESSAGE));

            return;
        }

        try {
            BigDecimal unitPrice = new BigDecimal(unitPriceStr);

            BigDecimal quantity = new BigDecimal(quantityStr);

            // Check if it has any modification
            if (purchase.getWalletTransaction().getWallet().getId().equals(wallet.getId())
                    && purchase.getWalletTransaction().getDescription().equals(description)
                    && purchase.getWalletTransaction().getStatus().equals(status)
                    && purchase.getWalletTransaction()
                            .getCategory()
                            .getId()
                            .equals(category.getId())
                    && purchase.getUnitPrice().compareTo(unitPrice) == 0
                    && purchase.getQuantity().compareTo(quantity) == 0
                    && purchase.getWalletTransaction().getDate().toLocalDate().equals(buyDate)
                    && purchase.getWalletTransaction().getIncludeInAnalysis()
                            == includeInAnalysisCheckBox.isSelected()) {
                WindowUtils.showInformationDialog(
                        preferencesService.translate(
                                Constants.TranslationKeys.INVESTMENT_DIALOG_NO_CHANGES_TITLE),
                        preferencesService.translate(
                                Constants.TranslationKeys
                                        .INVESTMENT_DIALOG_NO_CHANGES_PURCHASE_MESSAGE));
            } else // If there is any modification, update the transaction
            {
                LocalTime currentTime = LocalTime.now();
                LocalDateTime dateTimeWithCurrentHour = buyDate.atTime(currentTime);

                purchase.getWalletTransaction().setWallet(wallet);
                purchase.getWalletTransaction().setDescription(description);
                purchase.getWalletTransaction().setStatus(status);
                purchase.getWalletTransaction().setCategory(category);
                purchase.setUnitPrice(unitPrice);
                purchase.setQuantity(quantity);
                purchase.getWalletTransaction().setDate(dateTimeWithCurrentHour);
                purchase.getWalletTransaction()
                        .setIncludeInAnalysis(includeInAnalysisCheckBox.isSelected());

                tickerService.updatePurchase(purchase);

                WindowUtils.showSuccessDialog(
                        preferencesService.translate(
                                Constants.TranslationKeys.INVESTMENT_DIALOG_PURCHASE_UPDATED_TITLE),
                        preferencesService.translate(
                                Constants.TranslationKeys
                                        .INVESTMENT_DIALOG_PURCHASE_UPDATED_MESSAGE));
            }

            Stage stage = (Stage) tickerNameLabel.getScene().getWindow();
            stage.close();
        } catch (NumberFormatException e) {
            WindowUtils.showErrorDialog(
                    preferencesService.translate(
                            Constants.TranslationKeys.INVESTMENT_DIALOG_INVALID_NUMBER_TITLE),
                    preferencesService.translate(
                            Constants.TranslationKeys.INVESTMENT_DIALOG_INVALID_NUMBER_MESSAGE));
        } catch (EntityNotFoundException | IllegalArgumentException e) {
            WindowUtils.showErrorDialog(
                    preferencesService.translate(
                            Constants.TranslationKeys
                                    .INVESTMENT_DIALOG_ERROR_UPDATING_PURCHASE_TITLE),
                    e.getMessage());
        }
    }
}

File: src/main/kotlin/org/moinex/ui/dialog/investment/update/EditTickerPurchaseController.kt
/*
 * Filename: EditTickerPurchaseController.kt (original filename: EditTickerPurchaseController.java)
 * Created on: January 11, 2025
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 20/03/2026
 */

package org.moinex.ui.dialog.investment.update

import jakarta.persistence.EntityNotFoundException
import javafx.fxml.FXML
import javafx.stage.Stage
import org.moinex.common.constant.TranslationKeys
import org.moinex.common.extension.isEqual
import org.moinex.common.util.UIUtils
import org.moinex.common.util.WindowUtils
import org.moinex.model.enums.WalletTransactionType
import org.moinex.model.investment.TickerPurchase
import org.moinex.service.CategoryService
import org.moinex.service.PreferencesService
import org.moinex.service.investment.TickerService
import org.moinex.service.wallet.WalletService
import org.moinex.ui.dialog.investment.base.BaseTickerTransactionManagement
import org.springframework.stereotype.Controller
import java.time.LocalTime

@Controller
class EditTickerPurchaseController(
    walletService: WalletService,
    categoryService: CategoryService,
    tickerService: TickerService,
    preferencesService: PreferencesService,
) : BaseTickerTransactionManagement(
        walletService,
        categoryService,
        tickerService,
        preferencesService,
    ) {
    private lateinit var purchase: TickerPurchase

    init {
        walletTransactionType = WalletTransactionType.EXPENSE
    }

    fun setPurchase(p: TickerPurchase) {
        purchase = p
        tickerNameLabel.text = "${purchase.ticker.name} (${purchase.ticker.symbol})"
        unitPriceField.text = purchase.ticker.currentUnitValue.toString()

        setWalletComboBox(purchase.walletTransaction!!.wallet)

        descriptionField.text = purchase.walletTransaction!!.description
        unitPriceField.text = purchase.unitPrice.toString()
        quantityField.text = purchase.quantity.toString()
        statusComboBox.value = purchase.walletTransaction!!.status
        categoryComboBox.value = purchase.walletTransaction!!.category
        transactionDatePicker.value = purchase.walletTransaction!!.date.toLocalDate()
        includeInAnalysisCheckBox.isSelected = purchase.walletTransaction!!.includeInAnalysis

        totalPriceLabel.text = UIUtils.formatCurrency(purchase.walletTransaction!!.amount)
    }

    @FXML
    override fun handleSave() {
        val wallet = walletComboBox.value
        val description = descriptionField.text
        val status = statusComboBox.value
        val category = categoryComboBox.value
        val unitPriceStr = unitPriceField.text
        val quantityStr = quantityField.text
        val buyDate = transactionDatePicker.value

        if (wallet == null ||
            description.isNullOrBlank() ||
            status == null ||
            category == null ||
            unitPriceStr.isNullOrBlank() ||
            quantityStr.isNullOrBlank() ||
            buyDate == null
        ) {
            WindowUtils.showInformationDialog(
                preferencesService.translate(TranslationKeys.INVESTMENT_DIALOG_EMPTY_FIELDS_TITLE),
                preferencesService.translate(TranslationKeys.INVESTMENT_DIALOG_EMPTY_FIELDS_MESSAGE),
            )
            return
        }

        runCatching {
            val unitPrice = unitPriceStr.toBigDecimal()
            val quantity = quantityStr.toBigDecimal()

            if (purchase.walletTransaction!!.wallet.id == wallet.id &&
                purchase.walletTransaction!!.description == description &&
                purchase.walletTransaction!!.status == status &&
                purchase.walletTransaction!!.category.id == category.id &&
                purchase.unitPrice.isEqual(unitPrice) &&
                purchase.quantity.isEqual(quantity) &&
                purchase.walletTransaction!!.date.toLocalDate() == buyDate &&
                purchase.walletTransaction!!.includeInAnalysis == includeInAnalysisCheckBox.isSelected
            ) {
                WindowUtils.showInformationDialog(
                    preferencesService.translate(TranslationKeys.INVESTMENT_DIALOG_NO_CHANGES_TITLE),
                    preferencesService.translate(TranslationKeys.INVESTMENT_DIALOG_NO_CHANGES_PURCHASE_MESSAGE),
                )
            } else {
                val currentTime = LocalTime.now()
                val dateTimeWithCurrentHour = buyDate.atTime(currentTime)

                purchase.walletTransaction!!.wallet = wallet
                purchase.walletTransaction!!.description = description
                purchase.walletTransaction!!.status = status
                purchase.walletTransaction!!.category = category
                purchase.unitPrice = unitPrice
                purchase.quantity = quantity
                purchase.walletTransaction!!.date = dateTimeWithCurrentHour
                purchase.walletTransaction!!.includeInAnalysis = includeInAnalysisCheckBox.isSelected

                tickerService.updateTickerPurchase(purchase)

                WindowUtils.showSuccessDialog(
                    preferencesService.translate(TranslationKeys.INVESTMENT_DIALOG_PURCHASE_UPDATED_TITLE),
                    preferencesService.translate(TranslationKeys.INVESTMENT_DIALOG_PURCHASE_UPDATED_MESSAGE),
                )
            }

            (tickerNameLabel.scene.window as Stage).close()
        }.onFailure { e ->
            when (e) {
                is NumberFormatException -> {
                    WindowUtils.showErrorDialog(
                        preferencesService.translate(TranslationKeys.INVESTMENT_DIALOG_INVALID_NUMBER_TITLE),
                        preferencesService.translate(TranslationKeys.INVESTMENT_DIALOG_INVALID_NUMBER_MESSAGE),
                    )
                }
                is EntityNotFoundException, is IllegalArgumentException -> {
                    WindowUtils.showErrorDialog(
                        preferencesService.translate(TranslationKeys.INVESTMENT_DIALOG_ERROR_UPDATING_PURCHASE_TITLE),
                        e.message ?: "",
                    )
                }
                else -> throw e
            }
        }
    }
}
